### PR TITLE
Problems: free of stack variable, TODO left to evaluate

### LINKED ISSUE #2701: pgm build via cmake doesn't link libzmq with libpgm
steps to reproduce on linux

```
mkdir -p build && cd build && cmake -DWITH_OPENPGM=ON .. && make
sudo ldd lib/libzmq.so
```

The listing of linked libraries doesn't have the libpgm.

The build via `./autogen.sh && ./configure --with-pgm` works fine

### PR DESCRIPTION
**Pull request by bluca**
Solutions: see commits

Fixes #2701 #2699

### CODE REVIEW

**Comment by sigiesec** [comment]
@bluca, thanks for correcting my mistake in the call to free. It is quite hard to modify the source without having a chance to run or even compile it.

**Comment by bluca** [comment]
No problem!

**Comment by bluca** [comment]
@sigiesec if you are on Linux, with the changes in this PR you can install the library and build with it:

```
sudo apt install libkrb5-dev
./configure .... --with-libgssapi_krb5=yes
```

It won't run the test unless you setup the whole kerberos thing, but at least you can check it builds.

I'll see if I can add it by default to the packages builds on OBS, and then on a couple of CI jobs as well that already install packages.

File: tests/test_security_gssapi.cpp [hunk 1]
```diff
@@ -179,19 +185,23 @@ void test_unauth_creds (void *ctx, void *server, void *server_mon, char *endpoin
...
     rc = zmq_connect (client, endpoint);
     assert (rc == 0);
 
     expect_bounce_fail (server, client);
     close_zero_linger (client);
 
+#ifdef ZMQ_BUILD_DRAFT_API
     int event = get_monitor_event (server_mon, NULL, NULL);
-    assert (event == ZMQ_EVENT_HANDSHAKE_FAILED);
+    assert (event == ZMQ_EVENT_HANDSHAKE_FAILED_AUTH);
```
Comment (bluca) [inline tests/test_security_gssapi.cpp]: @sigiesec I took a very wild guess here and in the next events, is this correct?

**Comment by bluca** [comment]
There's a few errors in the GSSAPI code with older GCC on Debian 7, Ubuntu 12.04 and RHEL 7. I'll have a look during the weekend:

https://build.opensuse.org/package/live_build_log/home:bluca:branches:network:messaging:zeromq:git-draft/libzmq/Debian_7.0/x86_64

File: tests/test_security_gssapi.cpp (hunk 1 above)
Comment (sigiesec) [inline tests/test_security_gssapi.cpp]: This one looks good.

File: tests/test_security_gssapi.cpp [hunk 2]
```diff
@@ -205,8 +215,10 @@ void test_null_creds (void *ctx, void *server, void *server_mon, char *endpoint)
...
```
Comment (sigiesec) [inline tests/test_security_gssapi.cpp]: This should be ZMQ_EVENT_HANDSHAKE_FAILED_PROTOCOL with value ZMQ_PROTOCOL_ERROR_ZMTP_MECHANISM_MISMATCH  as in https://github.com/zeromq/libzmq/blob/7283574ceedbd4569f514d8a8a86117a0efbf65f/tests/test_security_curve.cpp#L178

Maybe expect_zmtp_mechanism_mismatch should be moved to testutil_security.hpp and reused here (and for the PLAIN client test as well)

File: tests/test_security_gssapi.cpp [hunk 3]
```diff
@@ -292,23 +304,30 @@ int main (void)
...
     rc = zmq_bind (server, "tcp://127.0.0.1:*");
     assert (rc == 0);
     rc = zmq_getsockopt (server, ZMQ_LAST_ENDPOINT, my_endpoint, &len);
     assert (rc == 0);
 
+#ifdef ZMQ_BUILD_DRAFT_API
     //  Monitor handshake events on the server
     rc = zmq_socket_monitor (server, "inproc://monitor-server",
-            ZMQ_EVENT_HANDSHAKE_SUCCEED | ZMQ_EVENT_HANDSHAKE_FAILED);
+            ZMQ_EVENT_HANDSHAKE_SUCCEEDED | ZMQ_EVENT_HANDSHAKE_FAILED_AUTH);
```
Comment (sigiesec) [inline tests/test_security_gssapi.cpp]: Here, ZMQ_EVENT_HANDSHAKE_FAILED_PROTOCOL must also be enabled.

File: tests/test_security_gssapi.cpp (hunk 2 above)
Comment (bluca) [inline tests/test_security_gssapi.cpp]: Thanks, sent a quick fix here: https://github.com/zeromq/libzmq/pull/2706

**Pull request merged**